File: wipl.py
# import numpy as np
import sys
from functools import lru_cache

sys.setrecursionlimit(10**6)
dp = [[0]*4020 for i in range(4020)]
pref = [0]*4020

# dp  = np.zeros((4020, 4020), dtype=int)
# pref = np.zeros(4020, dtype=int)

# @lru_cache
def recur(idx, taken, n, k, arr):
	if taken >= k and (pref[idx]-taken >= k):
		return 0
	elif idx >= n:
		return int(1e10)
	elif dp[idx][taken] != -1:
		return dp[idx][taken]

	c1 = recur(idx+1, min(taken+arr[idx], pref[idx]-taken), n, k, arr)
	c2 = recur(idx+1, min(pref[idx]-taken+arr[idx], taken), n, k, arr)

	dp[idx][taken] = 1+min(c1, c2)
	return dp[idx][taken]

def dp_solve(arr, n, k):
	for i in range(n//2):
		arr[i], arr[n-i-1] = arr[n-i-1], arr[i]
	for i in range(n+11):
		pref[i] = 0
		for j in range(k+11):
			dp[i][j] = -1
	for i in range(1, n+1):
		pref[i] = pref[i-1]+arr[i-1]
	ans = recur(0, 0, n, k, arr)
	if ans>int(1e9):
		return -1
	else:
		return ans


for _ in range(int(input())):
	n, k = map(int, input().split())
	arr = list(map(int, input().split()))
	arr.sort()
	# arr = tuple(arr)
	if n == 1:
		print(-1)
	else:
		print(dp_solve(arr, n, k))

File: wipl.cpp
#include <bits/stdc++.h>
using namespace std;


int dp[4010][4010];
int box(int *arr, int buf1, int buf2, int n) {
    if (buf1 <= 0 && buf2 <= 0)
        return 0;
    if (n < 0)
        return 6000;

    if (buf1 <= 0 || buf2 <= 0) {
        if (buf1 <= 0) {
            if (dp[0][buf2] > 0)
                return dp[0][buf2];
            int x = 1 + box(arr, buf1, buf2 - arr[n], n - 1);
            dp[0][buf2] = x;
            return x;
        }
        if (buf2 <= 0) {
            if (dp[buf1][0] > 0)
                return dp[buf1][0];
            int x = 1 + box(arr, buf1 - arr[n], buf2, n - 1);
            dp[buf1][0] = x;
            return x;
        }
    }
    if (dp[buf1][buf2] > 0)
        return dp[buf1][buf2];
    int x = min(1 + box(arr, buf1 - arr[n], buf2, n - 1),
                1 + box(arr, buf1, buf2 - arr[n], n - 1));
    dp[buf1][buf2] = x;
    return x;
}
void solve() {
    long long int n, k;
    cin >> n >> k;
    int arr[n];
    for (long long int i = 0; i < n; i++) {
        cin >> arr[i];
    }
    sort(arr, arr + n);
    for (int i = 0; i <= 4005; i++)
        for (int j = 0; j <= 4005; j++)
            dp[i][j] = -1;
    int x = box(arr, k, k, n - 1);
    if (x > 5000)
        cout << -1 << "\n";
    else
        cout << x << "\n";
}
int main() {
    int t;
    cin >> t;
    while (t--) {
        solve();
    }
}